obp/ope/estimators_embed.py

```diff
@@ -0,0 +1,666 @@
+        if position is None:
+            position = np.zeros(action_dist.shape[0], dtype=int)
+        if p_e_a is not None:
+            check_array(array=p_e_a, name="p_e_a", expected_dim=3)
+        else:
+            check_array(array=context, name="context", expected_dim=2)
+
+        if self.embedding_selection_method == "exact":
+            return self._estimate_with_exact_pruning(
+                context=context,
+                reward=reward,
+                action=action,
+                action_embed=action_embed,
+                position=position,
+                pi_b=pi_b,
+                action_dist=action_dist,
+            )
+
+        elif self.embedding_selection_method == "greedy":
+            return self._estimate_with_greedy_pruning(
+                context=context,
+                reward=reward,
+                action=action,
+                action_embed=action_embed,
+                position=position,
+                pi_b=pi_b,
+                action_dist=action_dist,
+            )
+
+        else:
+            return self._estimate_round_rewards(
+                context=context,
+                reward=reward,
+                action=action,
+                action_embed=action_embed,
+                position=position,
+                pi_b=pi_b,
+                action_dist=action_dist,
+                p_e_a=p_e_a,
+            ).mean()
+
+    def _estimate_with_exact_pruning(
+        self,
+        context: np.ndarray,
+        reward: np.ndarray,
+        action: np.ndarray,
+        action_embed: np.ndarray,
+        pi_b: np.ndarray,
+        action_dist: np.ndarray,
+        position: np.ndarray,
+    ) -> float:
+        """Apply an exact version of data-drive action embedding selection."""
+        n_emb_dim = action_embed.shape[1]
+        theta_list, cnf_list = [], []
+        feat_list, C = np.arange(n_emb_dim), np.sqrt(6) - 1
+        for i in np.arange(n_emb_dim, self.min_emb_dim - 1, -1):
+            comb_list = list(itertools.combinations(feat_list, i))
+            theta_list_, cnf_list_ = [], []
+            for comb in comb_list:
+                theta, cnf = self._estimate_round_rewards(
+                    context=context,
+                    reward=reward,
+                    action=action,
+                    action_embed=action_embed[:, comb],
+                    pi_b=pi_b,
+                    action_dist=action_dist,
+                    position=position,
+                    with_dev=True,
+                )
+                if len(theta_list) > 0:
+                    theta_list_.append(theta), cnf_list_.append(cnf)
+                else:
+                    theta_list.append(theta), cnf_list.append(cnf)
+                    continue
+
+            idx_list = np.argsort(cnf_list_)[::-1]
+            for idx in idx_list:
+                theta_i, cnf_i = theta_list_[idx], cnf_list_[idx]
+                theta_j, cnf_j = np.array(theta_list), np.array(cnf_list)
```

Comment:
[ask]

When `len(theta_list) == 1`, this function could return `theta_list[0]` even though `Eq. (22)` does not hold.

Is this an intended behavior?

I wonder if the initial value of `theta_list` should be the maximum value of `cnt_list_` of index `i`.

Reply:
`len(theta_list) == 1` means that there is only the initial value in `theta_list` where the initial value is the estimated policy value of MIPS when we use all available feature dimensions. 

After the initial value is added to `theta_list`, there is the `continue` statement of python in the loop, so the function outputs `theta_list[0]` only when `Eq. (22)` is true in the next iteration. 

Is this correct?

Reply:
OK, I misunderstand the implementation!!

That might be correct and has no problem!

Reply:
@fullflu thanks!